### PR TITLE
VM: Implement EIP 2537 (BLS12-381 precompiles)

### LINKED ISSUE #738: [Berlin] EIP-2537: BLS12-381 Curve Operation Precompiles
Not to be underestimated, this is a pretty big one (likely).

Might be a way to follow the path of the [rustbn.js](https://github.com/ethereumjs/rustbn.js) library and do an analogue integration on the [ZCash implementation](https://github.com/zcash/librustzcash/tree/master/pairing).

Eventually it is also possible (not sure about the performance implications) to take a pure JS implementation from the list of suggestions in the [EIP specification](https://github.com/ethereum/EIPs/pull/2537/files#diff-629a64fd292614a3917c7fb9a32f9337R290).

Discussion around this should start relatively imminent so that we can have some decision here.

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Closes #738 

Package used: [Herumi's MCL-WASM](https://github.com/herumi/mcl-wasm).

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
So there's currently 1 precompile in it (out of 9): the G1ADD precompile. I added some comments and TODO's for things I want to add.

One of the problems I have myself with the current changed I made is that the precompiles are now either async or not-async. I don't think this is clean, so I'd love to know what your thoughts are. The alternative is to make all precompiles by default async (even if they are not) and that thus `PrecompileFunc` always returns a `Promise<ExecResult>`.

The reason the precompiles are async is that the MCL package needs to get initialized; this is async. If you then try to call any of the MCL functions the chance is that MCL itself is not initialized and thus throws.

CC @evertonfraga @s1na @holgerd77

**Comment by jochem-brouwer** [comment]
Woops. It appears we have an async `VM.init()` function. So I'll revert the function signatures of the precompiles back to the old ones. Only have to figure out how to pass the MCL package in a clean way to the precompile functions.

**Comment by holgerd77** [comment]
Hi @jochem-brouwer, great start, nice! 😄 

Short note: you can use the pre-defined labels here to indicate your PR status (e.g. "WIP" or "Ready for Review")

**Comment by jochem-brouwer** [comment]
Heya! I have decided to keep keep support for async precompile functions. I think this might get useful in the future. Since the precompile functions are not exposed, this should not be a problem (?).

I have added the `mcl` package directly in the VM. It is globally instantiated and every BLS-precompile will read this package directly from the VM. The `VM` itself is now passed as an attribute of the `PrecompileInput`.

Thoughts? @holgerd77 @s1na @cgewecke @evertonfraga

**Comment by jochem-brouwer** [comment]
Note to self: if a file where MCL is imported in throws, then for some reason MCL also throws an extremely verbose error which we should supress.

**Comment by jochem-brouwer** [comment]
Linking issues of the mcl-wasm package;

[9](https://github.com/herumi/mcl-wasm/issues/9)

[10](https://github.com/herumi/mcl-wasm/issues/10)

[11](https://github.com/herumi/mcl-wasm/issues/11)

[12](https://github.com/herumi/mcl-wasm/issues/12)

**Comment by holgerd77** [comment]
@jochem-brouwer was scratching my head a bit on this VM -> precompile thing (passing the VM as a parameter and instantiate MCL in the VM, but I understand the reasoning (so to state explicitly: you want to avoid the performance/memory penalty on re-instantiating MCL again and again on every precompile, correct me if I am wrong). We shouldn't do these kind of things too often - here it might be justified for the moment - otherwise we introduce too much non-easy-to-grasp complexity. Also have no other idea atm, at least not without starting a somewhat major precompile structure refactoring, and give this some statefulness by making `Precompiles` an object or whatever.

Anyhow - no need to totally get distracted by solving this - but if someone has an alternative idea please post here.

**Comment by jochem-brouwer** [comment]
@holgerd77 Yes this problem is a bit confusing because it actually is two problems:

1) We have 9 precompiles which all need access to the same library.

2) The library has an `init()` function which is async. In this PR support for async precompiles is added, otherwise there's this condition that the library itself is not yet initialized and the precompile thus throws.

Upon thinking about 1): it does make sense if `require` always returns the same instance of the loaded library (99% sure this is correct). In that case I can remove this pointer to the VM in the `PrecompileInput` type. The only thing which I don't like here is that I find that initializing this library should be in `VM.init()`. Otherwise, every precompile has to manually `await` the promise of initializing this library. I find this a bit ugly and I don't think the code of the precompile is the right way to do this.

So just to summarize the options:

1) Keep `VM` pointer in the `PrecompileInput` so we can initialize the `async` init function of the library in `VM.init()`. 

2) Remove the `VM` pointer in `PrecompileInput` and instead explicitly `await` the `init` function of the library in each call to any of the 9 precompiles.

(2 does of course not impact performance; I just find the location of awaiting this initialization a bit ugly)

Thoughts? @holgerd77 @ryanio @cgewecke @s1na

**Comment by jochem-brouwer** [comment]
@evertonfraga Do you see why `mcl-wasm` is not installing on the checks?

**Comment by evertonfraga** [comment]
@jochem-brouwer I see the cause: it's using the CI code from master and trying to use for your PR, instead of using the CI instructions from your own branch.

I fail to understand why, though. It must be some new feature from github actions, because only they have control over these things.

The jobs being run in this PR are already using the dependency cache, even though the code for it is not integrated into this branch.

Loading cache: [CI run](https://github.com/ethereumjs/ethereumjs-vm/pull/785/checks?check_run_id=811931604#step:4:9)
But your branch does not use cache in CI: [eip-2537 vm-build.yml](https://github.com/ethereumjs/ethereumjs-vm/blob/eip-2537/.github/workflows/vm-build.yml#L11)
Only master does: [master vm-build.yml](https://github.com/ethereumjs/ethereumjs-vm/blob/master/.github/workflows/vm-build.yml#L27)

**Comment by evertonfraga** [comment]
I'd recommend to update this branch with latest master, and make sure to commit and push your package-lock.json.

**Comment by jochem-brouwer** [comment]
Cool - now we got all precompiles working! Only thing left to add are some error checks (only G1 failing tests work at this point) and a minor amount of cleaning some things so the code is consistent.

**Comment by jochem-brouwer** [comment]
Very exciting news! All tests from the [test dir](https://github.com/matter-labs/eip1962/tree/master/src/test/test_vectors/eip2537) pass locally.

However, I am having a discussion about [this particular test](https://raw.githubusercontent.com/matter-labs/eip1962/master/src/test/test_vectors/eip2537/extras/g2_multiexp.csv) currently as I am not entirely sure if this test is correct.

To do: integrate the official state tests into ethereumjs-testing and then also run the tests on these.

**Comment by jochem-brouwer** [comment]
Browser tests are timing out. We might want to remove all these csv files anyways since they will be tested in the state tests if these are implemented in the ethereumjs-testing lib.

**Comment by jochem-brouwer** [comment]
`npm run build && npm run test:state -- --fork=Berlin --dir=stEIP2537`

```
1..1403
# tests 1403
# pass  1403
```

😄 . (Running on a local fork of the tests repository, as currently `ethereumjs-testing` does not support the newest tests yet)

**Comment by evertonfraga** [comment]
> `npm run test:state -- --fork=Berlin --dir=stEIP2537`

@jochem-brouwer does the command you mentioned fully replaces the `.csv` testing logic? If so, we should look into releasing a new version of testing, instead of optimizing the csv.

**Comment by jochem-brouwer** [comment]
@evertonfraga Yes these tests are basically those CSVs but now converted to state tests (at the time I started copying these CSV tests they were not yet available as state tests). I agree, we should update the testing repo (will do so asap) and instead run state tests on these rather than trying to optimize these CSVs (I'll remove them).

**Comment by jochem-brouwer** [comment]
I will keep this open for review as the code itself is not likely to change after I've removed these CSV tests (I might add some tests which are not in the state tests though).

**Comment by holgerd77** [comment]
@jochem-brouwer Great stuff that this is getting ready for review 👍  (it's not quite ready yet with some CI runs not passing and some problems (from the chat) with the `ethereumjs-testing` update, right? Please explicitly ping once this is solved.)

One thing: PR is pretty commit-heavy with 49 commits. Can you squash several commits into one (please do a local repo backup before 🙂 ) so that we not bloat repo commit history too much? There are several consecutive cases like "lint, lint, do this, do this again, lint,..." which should be easily squashable without loosing information. Although the PR is also quite code heavy it should be possible to get this down by 15-25 commits at least.

**Comment by jochem-brouwer** [comment]
@holgerd77 I'll do a cleanup of these commits after this weekend.

Meanwhile the problem remains that the `test:API:browser` fails to work. The reason is:

[this require](https://github.com/herumi/mcl-wasm/blob/a6256e5004518934f735e0707c996010bed4aca2/mcl.js#L910)

actually throws (`mcl_c384_256.js: file not found`); since this promise does not have a `reject` call this means the promise never resolves (but also never throws) so in the VM, [this line](https://github.com/ethereumjs/ethereumjs-vm/blob/e94b17de4a190e112be5c181dbe2b2e6b38a4a9c/packages/vm/lib/index.ts#L176) keeps hanging forever (in the browser test).

I tried today to change some stuff in `karma.conf.js` but to no avail. @ryanio said he thought that maybe if we make karma serve these files as static it might work - would be great if someone could take a look at this as we cannot merge this PR if it fails to work in browser 😓

**Comment by ryanio** [comment]
@jochem-brouwer that's interesting, so even in karma `isNodeJs` is evaluating to true?

it looks like `isNodeJs` is being determined by checking `if (typeof exports === 'object')` ([link](https://github.com/herumi/mcl-wasm/blob/c1e8f86a432b4204595b2cc2304a1354360ec64f/mcl.js#L3)). I haven't seen that before to understand if it works well in practice, or if it may change with our TS compiler config.

If it *is* simply looking to require `./mcl_c384_256.js`, I wonder if we can nudge TS to include it by adding to tsconfig.prod:

```
"include": [
  "lib/**/*.ts",
  "node_modules/mcl-wasm/mcl_c384_256.js"
]
```

**Comment by jochem-brouwer** [comment]
@ryanio Tried adding this to `tsconfig` but this did not work.

I did figure out that `karma` has a `proxies` config which (with some tricks) seems to load the wasm file;

```
    proxies: {
      '/mcl_c384_256.wasm': "../../node_modules/mcl-wasm/mcl_c384_256.wasm",
    },
```

But it now throws inside the `fetch` block:

`ReferenceError: Module is not defined`

**Comment by evertonfraga** [comment]
The generated JS bundle creates a global `exports` object, so this is messing with the conditions mentioned above. Will keep investigating, but for a start, MCL needs better platform detection.

image

```
  if (typeof exports === 'object') {
    const crypto = require('crypto')
    crypto.getRandomValues = crypto.randomFillSync
    generator(exports, crypto, true)
  } else {
    const crypto = window.crypto || window.msCrypto
    const exports = {}
    window.mcl = generator(exports, crypto, false)
  }
```
=> always true in our context.

**Comment by evertonfraga** [comment]
**Problem 2) `Module()`**
I am really not sure where the `Module()` comes from. No other reference in the module itself or coming from any dependency. So I assumed it was just a mistake and did something similar to the node code path:

image

File: packages/vm/tests/tester.js
Comment (evertonfraga) [inline packages/vm/tests/tester.js:100]:
```suggestion
      const common = new Common('mainnet', FORK_CONFIG_VM)
```

File: packages/vm/tests/api/state/stateManager.js
Comment (evertonfraga) [inline packages/vm/tests/api/state/stateManager.js:165]: I am a bit confused to see an `async` refactor in this PR. Was it ported from another branch, or the changes were made here indeed? If the latter, it becomes harder to review when there are changes from  different scopes.

Also, is this test still being skipped on browsers?

Comment (jochem-brouwer) [reply packages/vm/tests/api/state/stateManager.js:165]: This is not changed here but is probably the result of a rebase which went wrong. I don't know how to fix it, but for some reason at one of the rebases it shows all changes from master as newly added/deleted items 😢

**Comment by jochem-brouwer** [comment]
Sat down and did first checked every commit in order to cleanup the changelog, then rebased `master`. Looks like I did a `git merge` at some point instead of `git rebase` which clogged up the commit logs. Also squashed the commits to a healthy amount of 13 (currently).

As we have discussed we will (temporarily) disable BLS precompiles in browser until we figure out what is causing these problems on the CI.

**Comment by jochem-brouwer** [comment]
I am removing the BLS EIP from `Berlin` HF and implementing the VM EIP support.

Some notes:

(1) as we have discussed we should move the EIP support to Common. 

(2) we should move precompile availability checks into Common. `precompiles/index.ts` is now getting a bit bloated to check this availability, and it would make a lot of sense that this support is in Common. 

If at some point we implement an EIP which wants to create a precompile at one of the addresses which is already occupied (like an EIP which wants to use a BLS address) then we can check in the Common (1) if there is a precompile and (2) what it is, as it might not be the case that "precompile is available" means that it is always the same one.

**Comment by jochem-brouwer** [comment]
Tests should now pass, the ones which run EIP2537 are disabled in browser now. Also note that I removed some tests from the PR CI (Dao and Homestead). Opened an issue regarding that BLS does not run in browser.

File: packages/vm/README.md
Comment (holgerd77) [inline packages/vm/README.md:66]: Won't make this a blocker but please do not update `README` files along feature PRs yet. Let's please keep this to the release PRs once the releases are actually done. Then one can go through the - now already existing 😄  - release notes and do a consistent update on the `README` files. This will otherwise confuse users of the latest releases published on npm.

File: packages/vm/lib/evm/precompiles/index.ts
Comment (holgerd77) [inline packages/vm/lib/evm/precompiles/index.ts:135]: Note: this new `vm` parameter can be removed again once we have `EIP` support moved to `Common`.

**Review by holgerd77 — APPROVED** [review]
Updated the branch, tests pass, this looks good now, thanks @jochem, really impressive work! 😄 

I gave this only a structural review and mainly looked at the integration code parts, these look good. I also skimmed through the implementation of the single precompiles itself but just for a rough look. Since I don't think that we realistically get someone on the level to do an in-depth review of the curve logic here I think we can (and should, to be able to move forward) on the current state of the DRAFT EIP merge in here and open this up for experimentation. Official tests pass which gives a somewhat high (with some deficits as Jochem mentioned) level of assurance that the implementation is correct. If the implementation is given a final ok for Berlin integration we should nevertheless give this code another look and eventually also actively help (now with the new tutorial on test creation 😄 ) to add to the official test suite with more test cases.

One high-level note on the precompile implementation: on a structural level there is a high level of structural similarities - e.g. on the zero bytes check - I wonder if it's worth to add 1-3 util functions which can be shared and reduce the size of the code base here.

Will now merge this in. The EIP integration part will need some update on the shift to `Common`, but I can do that on the follow-up work on #863

File: packages/vm/README.md
Comment (jochem-brouwer) [reply packages/vm/README.md:66]: We do update the changelog right? So the changelog does have new versions (which are not available on npm yet, which is also a bit confusing), but we keep the readme at the latest published version?

File: packages/vm/lib/evm/precompiles/index.ts
Comment (jochem-brouwer) [reply packages/vm/lib/evm/precompiles/index.ts:135]: Yes, should fix once that is merged in

**Pull request merged**
